### PR TITLE
Private/gokay/clipboard 2

### PR DESCRIPTION
**Pull request by gokaysatir**
* Resolves: # 
* Target version: master 

### TODO

- [ ] ...

### Checklist

- [ ] I have run `make prettier-write` and formatted the code.
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

File: browser/src/control/Control.DownloadProgress.js [hunk 1]
```diff
@@ -114,12 +111,14 @@ L.Control.DownloadProgress = L.Control.extend({
 		var buttonText = _('Copy') + L.Util.replaceCtrlAltInMac(' (Ctrl + C)');
 
 		if (inSnackbar) {
-			this._map.uiManager.showProgressBar(snackbarMsg, buttonText,
-				this._onConfirmCopyAction.bind(this), this.options.snackbarTimeout);
-
 			this._map.uiManager.setSnackbarProgress(100);
 
-			this.setupKeyboardShortcutForSnackbar();
+			this._map.uiManager.showSnackbar(
```
Comment (vmiklos) [inline browser/src/control/Control.DownloadProgress.js:116]: You also replace the progressbar with a snackbar here, which is not mentioned in the commit message. But it's probably OK. Could you please describe why do you change it, though?

File: browser/src/map/Clipboard.js [hunk 2]
```diff
@@ -1170,7 +1170,7 @@ L.Clipboard = L.Class.extend({
 		if (!this._downloadProgress || this._downloadProgress.isClosed())
 			return;
 
-		if (this._downloadProgressStatus() === 'downloadButton')
+		if (['downloadButton', 'confirmPasteButton'].includes(this._downloadProgressStatus()))
```
Comment (vmiklos) [inline browser/src/map/Clipboard.js:1173]: Same here: OK to hide the download widget in more cases, but could you describe why is this necessary? What is the case when we get here with a downloadButton case and when do we get confirmPasteButton?

File: browser/src/control/Control.DownloadProgress.js [hunk 3]
```diff
@@ -150,7 +153,7 @@ L.Control.DownloadProgress = L.Control.extend({
 	},
 
 	setupKeyboardShortcutForSnackbar: function () {
-		this._setupKeyboardShortcutForElement('snackbar', 'button');
+		this._setupKeyboardShortcutForElement('snackbar-container', 'button');
```
Comment (vmiklos) [inline browser/src/control/Control.DownloadProgress.js:155]: OK, but why is this needed? Not clear how we had snackbar here in the past and how snackbar-container is better now.

File: browser/src/control/Control.DownloadProgress.js [hunk 4]
```diff
@@ -64,6 +64,9 @@ L.Control.DownloadProgress = L.Control.extend({
 				msg, buttonText, this._onStartDownload.bind(this), this.options.snackbarTimeout);
 
 			this.setupKeyboardShortcutForSnackbar();
+
+			// Snackbars cannot get focus, but we are using it with a button and it requires to be focused. No need to change snackbar imp for now.
+			document.getElementById('button').focus(); // TODO: This "button" id is too generic. It could be something like "snackbar-button" etc.
```
Comment (vmiklos) [inline browser/src/control/Control.DownloadProgress.js]: https://github.com/CollaboraOnline/online/pull/10394/commits/213974e6619b55010008d3aba17f5277201d0f42 has a check here for the case when the getElementById() result is undefined, is it intentional that you don't do the same here? Maybe extract to a function to avoid the duplication?

File: browser/src/control/Control.JSDialog.js [hunk 5]
```diff
@@ -400,7 +400,7 @@ L.Control.JSDialog = L.Control.extend({
 			focusWidget = instance.init_focus_id ? instance.container.querySelector('[id=\'' + instance.init_focus_id + '\']') : null;
 		}
 
-		if (focusWidget && document.activeElement !== focusWidget) {
+		if (focusWidget && document.activeElement !== focusWidget && instance.canHaveFocus) {
```
Comment (vmiklos) [inline browser/src/control/Control.JSDialog.js:403]: Is it fine to just not do anything in this case? You don't mention in a comment or in the commit message if you assume that the outer element has the focus already at this point.

File: browser/src/map/Clipboard.js [hunk 6]
```diff
@@ -536,33 +530,31 @@ L.Clipboard = L.Class.extend({
 	populateClipboard: function(ev) {
 		this._checkSelection();
 
-		if (this._navigatorClipboardWrite()) {
-			// This is the codepath where the browser initiates the clipboard operation,
```
Comment (vmiklos) [inline browser/src/map/Clipboard.js:540]: I found this comment useful: there are two cases to handle (user clicks on copy button or user presses Ctrl-C) and it's not trivial which one is handled where. Would you consider restoring it?

Also, the early return had the benefit of keeping the intent level small. A usual refactor is to "flatten" a function by adding such early returns. Is there a benefit to do the opposite of that here?

File: browser/src/map/Clipboard.js [hunk 7]
```diff
@@ -501,12 +501,6 @@ L.Clipboard = L.Class.extend({
 	_getHtmlForClipboard: function() {
 		var text;
 
-		if ($('.ui-edit').is(':focus'))
-			return $('.ui-edit').value();
-
-		if ($('.w2ui-input').is(':focus'))
-			return $('.w2ui-input').value();
-
```
Comment (vmiklos) [inline browser/src/map/Clipboard.js:509]: Did you test if copying from comments still work? That kind of info would go to the commit message ideally. :-)

File: browser/src/map/Clipboard.js [hunk 8]
```diff
@@ -857,10 +857,6 @@ L.Clipboard = L.Class.extend({
 	// Executes the navigator.clipboard.write() call, if it's available.
 	_navigatorClipboardWrite: function() {
 		if (!L.Browser.clipboardApiAvailable) {
-			// Show a visible warning, this should not happen in production.
-			this._map.uiManager.showSnackbar(
-				_('The async Clipboard API is not supported by your browser, switching to HTTPS is meant to fix that.')
-			);
```
Comment (vmiklos) [inline browser/src/map/Clipboard.js:863]: OK. I think initially @eszkadev wanted this, because first it was not clear that http:// gives no new/async clipboard, even on the latest Chrome. But if the warning causes more problems than it solves, then let's forget about it.

File: browser/src/map/Clipboard.js [hunk 9]
```diff
@@ -923,7 +923,7 @@ L.Clipboard = L.Class.extend({
 			L.Browser.clipboardApiAvailable = false;
 			window.prefs.set('clipboardApiAvailable', 'false');
 			// Prefetch selection, so next time copy will work with the keyboard.
-			app.socket.sendMessage('gettextselection mimetype=text/html,text/plain;charset=utf-8');
+			app.socket.sendMessage(that.getMetaURL() + '&MimeType=text/html,text/plain;charset=utf-8');
```
Comment (vmiklos) [inline browser/src/map/Clipboard.js:914]: This one is really strange: the socket messages typically have a syntax where a lowercase command starts the line, then we have various parameters here. Do you really send a raw URL here or do I misread this change?

Also, you don't really describe what is the problem you try to fix here. :-)

File: wsd/ClientRequestDispatcher.cpp [hunk 10]
```diff
@@ -1113,8 +1114,13 @@ bool ClientRequestDispatcher::handleClipboardRequest(const Poco::Net::HTTPReques
             tag = it.second;
         else if (it.first == "MimeType")
             mime = it.second;
+        else if (it.first == "charset")
+            charset = it.second;
     }
 
+    if (!charset.empty())
+        mime += ";charset=" + charset;
```
Comment (vmiklos) [inline wsd/ClientRequestDispatcher.cpp:1122]: I assume that the charset is then parsed by core, since you only produce this charset info, you don't add code to handle it somewhere. If this is the case, we could have a comment here stating this.

**Review by vmiklos — CHANGES_REQUESTED** [review]
I added a number of inline comments with nits, but those are just small things, this mostly looks OK to me.

One problematic change stands out, though: could you please avoid `sendMessage(that.getMetaURL()...)` e.g. by keeping the gettextselection command there? Thanks. The rest looks good to me.

File: wsd/ClientRequestDispatcher.cpp (hunk 10 above)
Comment (gokaysatir) [reply wsd/ClientRequestDispatcher.cpp:1122]: It's handled below in line ~1167:

` else if (mime == "text/html,text/plain;charset=utf-8")`

File: browser/src/control/Control.JSDialog.js (hunk 5 above)
Comment (gokaysatir) [reply browser/src/control/Control.JSDialog.js:403]: I think if "`instance.canHaveFocus`" is false then we musn't check for focusables. But maybe @eszkadev has a better idea here? This part may need some refactoring.

Comment (gokaysatir) [reply browser/src/control/Control.JSDialog.js:403]: Because that if statement conflicts with `instance.canHaveFocus` property.

File: browser/src/control/Control.DownloadProgress.js (hunk 3 above)
Comment (gokaysatir) [reply browser/src/control/Control.DownloadProgress.js:155]: Narrows down the search. Not vital though :)

File: browser/src/map/Clipboard.js (hunk 2 above)
Comment (gokaysatir) [reply browser/src/map/Clipboard.js:1173]: 'downloadButton' doesn't seem to be used, added to the commit message.

File: browser/src/control/Control.DownloadProgress.js (hunk 1 above)
Comment (gokaysatir) [reply browser/src/control/Control.DownloadProgress.js:116]: Added a note.

File: browser/src/map/Clipboard.js (hunk 6 above)
Comment (gokaysatir) [reply browser/src/map/Clipboard.js:544]: Restored :)

File: browser/src/map/Clipboard.js (hunk 9 above)
Comment (gokaysatir) [reply browser/src/map/Clipboard.js]: :D Forgotten debugging thing.

File: browser/src/map/Clipboard.js (hunk 7 above)
Comment (gokaysatir) [reply browser/src/map/Clipboard.js:513]: Yep, tested just before sending the changes.

**Comment by gokaysatir** [comment]
> I added a number of inline comments with nits, but those are just small things, this mostly looks OK to me.
> 
> One problematic change stands out, though: could you please avoid `sendMessage(that.getMetaURL()...)` e.g. by keeping the gettextselection command there? Thanks. The rest looks good to me.


Thanks for checking :) I sent the changes and tested them with traefik software befor sending.

File: browser/src/map/Clipboard.js [hunk 11]
```diff
@@ -334,13 +334,9 @@ L.Clipboard = L.Class.extend({
 				);
 			}.bind(this),
 			function(progress) { return progress/2; },
-			function(response) {
+			function() {
 				window.app.console.log('failed to download clipboard using fallback html');
 
-				if (response) {
```
Comment (vmiklos) [inline browser/src/map/Clipboard.js:340]: Nit: why did you remove this block? This is not something that should happen in practice, so it sounds useful to show a user-visible error in this case?

File: browser/src/map/Clipboard.js (hunk 9 above)
Comment (vmiklos) [reply browser/src/map/Clipboard.js]: I see this is gone from the latest version, good.

**Review by vmiklos — APPROVED** [review]
Looks good to me now, thanks.

In the future, please try to always have these paragraphs in commit messages:

- what is happening currently (as a user)
- problem description (what you think is the root of the problem, as a developer)

I hope that helps to describe the "why" in commit messages so I nag you less. :-) Thanks.

File: browser/src/map/Clipboard.js (hunk 11 above)
Comment (gokaysatir) [reply browser/src/map/Clipboard.js:340]: There is a warning via "showInfoModal" few lines below. Looks like it will be shown when the operation fails. So i removed the first one.

Comment (vmiklos) [reply browser/src/map/Clipboard.js:340]: I see, thanks.
